core/provider/docker/provider.go

```diff
@@ -0,0 +1,358 @@
+		return nil, err
+	}
+
+	portSet := convertTaskDefinitionPortsToPortSet(definition)
+	portBindings, err := p.GeneratePortBindings(portSet)
+	if err != nil {
+		return nil, fmt.Errorf("failed to allocate task ports: %v", err)
+	}
+
+	var mounts []mount.Mount
+
+	logger.Debug("creating task", zap.String("name", definition.Name), zap.String("image", definition.Image.Image))
+
+	if definition.DataDir != "" {
+		volumeName := fmt.Sprintf("%s-data", definition.Name)
+
+		logger.Debug("creating volume", zap.String("name", volumeName))
+
+		_, err = p.CreateVolume(ctx, provider.VolumeDefinition{
+			Name:      volumeName,
+			Size:      "10GB",
+			MountPath: definition.DataDir,
+		})
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to create dataDir: %v", err)
+		}
+
+		volumeMount := mount.Mount{
+			Type:   mount.TypeVolume,
+			Source: volumeName,
+			Target: definition.DataDir,
+		}
+
+		logger.Debug("setting volume owner", zap.String("name", volumeName), zap.String("uid", definition.Image.UID), zap.String("gid", definition.Image.GID))
+
+		if err = p.SetVolumeOwner(ctx, volumeName, definition.Image.UID, definition.Image.GID); err != nil {
+			return nil, fmt.Errorf("failed to set volume owner: %v", err)
+		}
+
+		mounts = []mount.Mount{volumeMount}
+
+		taskState.Volume = &VolumeState{
+			Name: volumeName,
+			Size: "10GB",
+		}
+	}
+
+	logger.Debug("creating container", zap.String("name", definition.Name), zap.String("image", definition.Image.Image))
+
+	// network map is volatile, so we need to mutex update it
+	ip, err := p.nextAvailableIp()
+	if err != nil {
+		return nil, err
+	}
+
+	createdContainer, err := p.dockerClient.ContainerCreate(ctx, &container.Config{
+		Image:      definition.Image.Image,
+		Entrypoint: definition.Entrypoint,
+		Cmd:        definition.Command,
+		Tty:        false,
+		Hostname:   definition.Name,
+		Labels: map[string]string{
+			providerLabelName: p.state.Name,
+		},
+		Env: convertEnvMapToList(definition.Environment),
+	}, &container.HostConfig{
+		Mounts:          mounts,
+		PortBindings:    portBindings,
+		PublishAllPorts: true,
+		NetworkMode:     container.NetworkMode(p.state.NetworkName),
+	}, &network.NetworkingConfig{
+		EndpointsConfig: map[string]*network.EndpointSettings{
+			p.state.NetworkName: {
+				IPAMConfig: &network.EndpointIPAMConfig{
+					IPv4Address: ip,
+				},
+			},
+		},
+	}, nil, definition.ContainerName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	taskState.Id = createdContainer.ID
+	taskState.Status = provider.TASK_STOPPED
+
+	p.stateMu.Lock()
+	defer p.stateMu.Unlock()
+
+	p.state.TaskStates[taskState.Id] = taskState
+
+	return &Task{
+		state:    taskState,
+		provider: p,
+	}, nil
+}
+
+func (p *Provider) SerializeProvider(context.Context) ([]byte, error) {
+	p.stateMu.Lock()
+	defer p.stateMu.Unlock()
+
+	bz, err := json.Marshal(p.state)
+
+	return bz, err
+}
+
+func (p *Provider) SerializeTask(ctx context.Context, task provider.TaskI) ([]byte, error) {
```

Comment:
Serial and Deserialize Task functions do not need to be receiver methods on `Provider`.

If we are always casting these interfaces to the concrete `Task` type, it might also make more sense to just accept the concrete type as the parameter instead of the interface

Reply:
we use locks here, so we they do need to be a receiver method, otherwise the mutex will be ineffective